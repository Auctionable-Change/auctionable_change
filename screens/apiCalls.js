const url = 'https://auctionable-change-api.herokuapp.com/items'

export const fetchItems = async () => {
  try {
    const response = await fetch(url);
    if(!response.ok) {
      throw new Error(error);
    }
    const data = await response.json();
    return data;
  } catch(error) {
    return false
  }
}

export const postItem = async (listing) => {
  const response = await fetch(url, {
    method: 'POST', 
    headers: {
      'Content-Type': 'application/json'
    },
    body: JSON.stringify(listing)
  })
  return response.json()
}


export const fetchCharities = async (search) => {
  try {
    const response = await fetch(`https://auctionable-change-api.herokuapp.com/charities/${search}`)
    if(!response.ok) {
      throw new Error(error);
    }
    const data = await response.json();
    return data;
  } catch(error) {
    return false
  }
}

let CLOUDINARY_URL = 'https://api.cloudinary.com/v1_1/djk5anakm/upload';


export const cloudinaryPost = (pickerResult, sendEmail) => {
  let photoData;
    let base64Img = `data:image/jpg;base64,${pickerResult.base64}`;
    let data = {
      file: base64Img,
      upload_preset: "oclsdjxp",
    };

    fetch(CLOUDINARY_URL, {
      body: JSON.stringify(data),
      headers: {
        "content-type": "application/json",
      },
      method: "POST",
    })
      .then(async (res) => await res.json())
      .then((res) => (photoData = res.url))
      .then(() => console.log("photo data", photoData))
      .then(() => sendEmail(photoData));
  };

