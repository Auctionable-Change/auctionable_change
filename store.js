import React, { createContext, useContext, useReducer } from "react";

const StoreContext = createContext();

const initialState = {
  currentListing: {
    id: null,
    title: "",
    description: "",
    price: null,
    donor: "",
    donor_email: "",
    status: "",
    category: "",
    charity: "",
    charity_url: "",
    charity_score: "",
    charity_score_image: "",
    image: "",
    bids: [],
  },
  buyerDetails: {
    item_id: null,
    bidder_name: "",
    bidder_email: "",
    amount: null,
    street_address: "",
    city: "",
    state: "",
    zip_code: "",
    receipt: "",
  },
  listingToPost: {},
};

const reducer = (state, action) => {
  switch (action.type) {
    case "ADD_CURRENT_LISTING":
      return {
        ...state,
        currentListing: action.currentListing,
      };
    case "ADD_SELECTED_CHARITY":
      return {
        ...state,
        listingToPost: {
          ...state.listingToPost,
          charity: action.selectedCharity.name,
          charity_url: action.selectedCharity.url,
          charity_score: action.selectedCharity.rating,
          charity_score_image: action.selectedCharity.rating_image,
        },
      };
    case "ADD_BUYER_DETAILS":
      return {
        ...state,
        buyerDetails: { ...state.buyerDetails, ...action.buyerDetails },
      };
    case "ADD_TO_LISTING":
      return {
        ...state,
        listingToPost: { ...state.listingToPost, ...action.listingToPost },
      };
    case "POST_BID":
      return {
        ...state,
        currentBid: action.bid
      };
      case "SET_USER":
      return {
        ...state,
        loginDetails: action.loginDetails,
        userInfo: action.userInfo,
        registrationDetails: action.registration
      }

    default:
      throw new Error(`Unhandled action type: ${action.type}`);
  }
};

export const StoreProvider = ({ children }) => {
  const [state, dispatch] = useReducer(reducer, initialState);

  return (
    <StoreContext.Provider value={{ state, dispatch }}>
      {children}
    </StoreContext.Provider>
  );
};

export const useStore = () => useContext(StoreContext);
