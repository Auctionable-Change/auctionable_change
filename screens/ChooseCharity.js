import * as React from "react";
import { SafeAreaView, Text } from "react-native";
import { StyleSheet } from "react-native";
import { TouchableOpacity, ScrollView } from "react-native-gesture-handler";
import { useStore } from "../store";

const ChooseCharity = ({ navigation }) => {
  return (
    <SafeAreaView style={styles.container}>
      <ScrollView>
        <Text>Choose Charity</Text>
        <TouchableOpacity
          onPress={() => navigation.navigate("Confirmation")}
          style={styles.button}
        >
          <Text>Confirm Post</Text>
        </TouchableOpacity>
      </ScrollView>
    </SafeAreaView>
  );
};

const styles = StyleSheet.create({
  container: {
    flex: 1,
    alignItems: "center",
    justifyContent: "center",
  },
  button: {
    backgroundColor: "#FFFFFF",
    borderRadius: 10,
    width: 200,
    padding: 10,
    margin: 10,
    alignItems: "center",
    shadowOffset: {
      width: 0,
      height: 1,
    },
    shadowOpacity: 0.18,
    shadowRadius: 1.0,
  },
});

export default ChooseCharity;
