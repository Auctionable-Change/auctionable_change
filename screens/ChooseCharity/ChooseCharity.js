import React, { useState } from "react";
import { View, Image } from "react-native";
import { SafeAreaView } from "react-native-safe-area-context";
import { StyleSheet, Linking, ActivityIndicator } from "react-native";
import { FlatList } from "react-native-gesture-handler";
import { useStore } from "../../store";
import { fetchCharities } from "../apiCalls";
import {
  Icon,
  Card,
  CheckBox,
  Item,
  Input,
  Button,
  Text,
  Right,
} from "native-base";

const ChooseCharity = ({ navigation }) => {
  const { dispatch } = useStore();
  const [search, setSearch] = useState(null);
  const [isLoading, setIsLoading] = useState(false);
  const [charitySelected, setCharitySelected] = useState(false);
  const [charities, setCharities] = useState(null);

  const handleChange = (event) => {
    setSearch(event.nativeEvent.text);
  };

  const returnCharities = async () => {
    setCharities([]);
    setIsLoading(true);
    const data = await fetchCharities(search);
    setIsLoading(false);
    setCharities(data.charities);
  };

  const charitySelect = (name) => {
    let charity = charities.find((charity) => charity.name === name);
    setCharitySelected(charity);
  };

  const confirmCharity = () => {
    navigation.navigate("Confirmation");
    dispatch({
      type: "ADD_SELECTED_CHARITY",
      selectedCharity: charitySelected,
    });
  };

  return (
    <SafeAreaView style={{ flex: 1, alignItems: "center" }}>
      <View style={{ flexDirection: "row" }}>
        <Item style={{ width: "75%" }}>
          <Icon name="ios-search" />
          <Input
            style={{ fontFamily: "quicksand", fontSize: 15 }}
            placeholder="Search Charities"
            onChange={(event) => handleChange(event)}
          ></Input>
        </Item>
        <Button success onPress={() => returnCharities()}>
          <Text style={{ fontFamily: "quicksand-bold", fontSize: 15 }}>
            Go!
          </Text>
        </Button>
      </View>
      {!charities && (
        <Card style={{ width: "90%", height: 150, marginTop: 50 }}>
          <Text
            style={{
              alignSelf: "center",
              paddingTop: 30,
              fontFamily: "quicksand",
              fontSize: 15,
            }}
          >
            Search For Your Favorite Charity Above
          </Text>
          <Text
            style={{
              alignSelf: "center",
              margin: 5,
              fontFamily: "quicksand",
              fontSize: 15,
            }}
          >
            - or -
          </Text>
          <Button
            bordered
            success
            style={{ alignSelf: "center" }}
            onPress={() => returnCharities()}
          >
            <Text style={{ fontFamily: "quicksand", fontSize: 15 }}>
              Browse Our Featured Charities
            </Text>
          </Button>
        </Card>
      )}
      {isLoading && (
        <View style={{ flex: 1, justifyContent: "center" }}>
          <ActivityIndicator size="large" color="#2CB833" />
        </View>
      )}

      <FlatList
        data={charities}
        style={styles.scrollView}
        keyExtractor={(item) => item.name}
        renderItem={({ item }) => (
          <Card
            style={
              item.name == charitySelected.name
                ? styles.charityActive
                : styles.charityDefault
            }
          >
            <View>
              <Button
                transparent
                onPress={() => Linking.openURL(item.url)}
                style={{
                  height: "100%",
                  alignSelf: "center",
                  flex: 5,
                }}
              >
                <Text
                  style={{
                    alignSelf: "center",
                    fontSize: 18,
                    color: "#439fd9",
                    fontFamily: "quicksand",
                  }}
                  numberOfLines={2}
                >
                  {item.name}
                </Text>
              </Button>
              <Image
                source={{ uri: item.rating_image }}
                style={{
                  width: 80,
                  height: 40,
                  resizeMode: "contain",
                  marginLeft: 15,
                }}
              />
            </View>
            <Right>
              <CheckBox
                color="#2CB833"
                onPress={() => charitySelect(item.name)}
                checked={item.name == charitySelected.name ? true : false}
                style={{ margin: 15 }}
              />
            </Right>
          </Card>
        )}
      />

      {charitySelected && (
        <Button block success onPress={() => confirmCharity()}>
          <Text
            style={{
              color: "white",
              fontFamily: "quicksand-bold",
              fontSize: 15,
            }}
          >
            Confirm Charity & Continue
          </Text>
        </Button>
      )}
    </SafeAreaView>
  );
};

const styles = StyleSheet.create({
  search: {
    width: "100%",
    alignItems: "center",
    justifyContent: "center",
  },

  scrollView: {
    flex: 1,
    width: "90%",
  },
  charityDefault: {
    backgroundColor: "#FFF",
    padding: 15,
    width: "98%",
    marginTop: 15,
    height: 130,
    flexDirection: "row",
  },
  charityActive: {
    backgroundColor: "#FFF",
    borderWidth: 5,
    height: 130,
    borderColor: "blue",
    padding: 15,
    width: "98%",
    marginTop: 15,
    flexDirection: "row",
  },
});

export default ChooseCharity;
