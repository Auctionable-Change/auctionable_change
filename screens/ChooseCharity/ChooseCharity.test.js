import React from 'react';
import { render, fireEvent, waitFor, NativeTestEvent } from 'react-native-testing-library';
import ChooseCharity from './ChooseCharity';
import { StoreProvider } from '../../store'
import {expect, it} from '@jest/globals';
import renderer from "react-test-renderer";
import { fetchCharities } from '../apiCalls';

jest.mock('../apiCalls')

describe('ChooseCharity', () => {
 it('should render', () => {
  const tree = renderer.create(
    <StoreProvider>
      <ChooseCharity/>
    </StoreProvider>).toJSON();
  expect(tree).toMatchSnapshot();
 })
 it('should display a message before searching for a charity', () => {
   const { getByText } = render(
    <StoreProvider>
      <ChooseCharity/>
    </StoreProvider>)

    const message = getByText('Search For Your Favorite Charity Above')
    const button = getByText('Browse Our Featured Charities')
    expect(message).toBeTruthy()
    expect(button).toBeTruthy()
 })
 it('should search the BE charity API', () => {
    fetchCharities.mockResolvedValue(
      {
        "id": 200345249,
        "url": "https://www.charitynavigator.org/?bay=search.summary&orgid=14066&utm_source=DataAPI&utm_content=6bc9f3c8",
        "name": "Charity1",
        "rating": 4
      },
    )
    const { getByText, getByPlaceholder } = render(
      <StoreProvider>
        <ChooseCharity />
      </StoreProvider>)

      const searchBtn = getByText('Go!');
      fireEvent.press(searchBtn);
      const charityItem = waitFor(() => getByText('Charity1'));
      expect(charityItem).toBeTruthy();
  
    })
   
})
