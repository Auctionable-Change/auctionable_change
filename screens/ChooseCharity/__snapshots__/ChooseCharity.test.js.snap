// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ChooseCharity should render 1`] = `
<RNCSafeAreaView
  style={
    Object {
      "alignItems": "center",
      "backgroundColor": "#FFFFFF",
      "flex": 1,
    }
  }
>
  <View
    style={
      Object {
        "flexDirection": "row",
      }
    }
  >
    <View
      accessible={true}
      focusable={false}
      forwardedRef={[Function]}
      onClick={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        Object {
          "alignItems": "center",
          "backgroundColor": "transparent",
          "borderBottomWidth": 1,
          "borderColor": "#D9D5DC",
          "borderLeftWidth": 0,
          "borderRadius": undefined,
          "borderRightWidth": 0,
          "borderTopWidth": 0,
          "borderWidth": undefined,
          "flexDirection": "row",
          "marginLeft": 2,
          "opacity": 1,
          "width": "75%",
        }
      }
    >
      <Text />
      <TextInput
        allowFontScaling={true}
        editable={true}
        onChange={[Function]}
        placeholder="Search Charities"
        placeholderTextColor="#575757"
        rejectResponderTermination={true}
        style={
          Array [
            Object {
              "color": "#000",
              "flex": 1,
              "fontSize": 17,
              "height": 50,
              "paddingLeft": 5,
              "paddingRight": 5,
              "top": 1.5,
            },
            Object {
              "fontFamily": "quicksand",
              "fontSize": 15,
            },
          ]
        }
        underlineColorAndroid="rgba(0,0,0,0)"
      />
    </View>
    <View
      accessible={true}
      focusable={true}
      forwardedRef={[Function]}
      onClick={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        Object {
          "alignItems": "center",
          "alignSelf": "flex-start",
          "backgroundColor": "#2cb833",
          "borderBottomWidth": null,
          "borderColor": "#007aff",
          "borderLeftWidth": null,
          "borderRadius": 5,
          "borderRightWidth": null,
          "borderTopWidth": null,
          "borderWidth": undefined,
          "elevation": 2,
          "flexDirection": "row",
          "height": 45,
          "justifyContent": "space-between",
          "opacity": 1,
          "paddingBottom": 6,
          "paddingTop": 6,
          "shadowColor": undefined,
          "shadowOffset": undefined,
          "shadowOpacity": undefined,
          "shadowRadius": undefined,
        }
      }
    >
      <Text
        style={
          Array [
            Object {
              "backgroundColor": "transparent",
              "color": "#fff",
              "fontFamily": "System",
              "fontSize": 16.5,
              "marginLeft": 0,
              "marginRight": 0,
              "paddingLeft": 16,
              "paddingRight": 16,
            },
            Object {
              "fontFamily": "quicksand-bold",
              "fontSize": 15,
            },
          ]
        }
      >
        Go!
      </Text>
    </View>
  </View>
  <View
    style={
      Array [
        Object {
          "backgroundColor": "#fff",
          "borderBottomWidth": 0.5,
          "borderColor": "#ccc",
          "borderLeftWidth": 0.5,
          "borderRadius": 2,
          "borderRightWidth": 0.5,
          "borderTopWidth": 0.5,
          "elevation": 3,
          "flexWrap": "nowrap",
          "marginBottom": 5,
          "marginLeft": 2,
          "marginRight": 2,
          "marginTop": 5,
          "shadowColor": "#000",
          "shadowOffset": Object {
            "height": 2,
            "width": 0,
          },
          "shadowOpacity": 0.1,
          "shadowRadius": 1.5,
        },
        Object {
          "height": 150,
          "marginTop": 30,
          "width": "90%",
        },
      ]
    }
  >
    <Text
      style={
        Array [
          Object {
            "color": "#000",
            "fontFamily": "System",
            "fontSize": 16,
          },
          Object {
            "alignSelf": "center",
            "fontFamily": "quicksand",
            "fontSize": 15,
            "paddingTop": 30,
          },
        ]
      }
    >
      Search For Your Favorite Charity Above
    </Text>
    <Text
      style={
        Array [
          Object {
            "color": "#000",
            "fontFamily": "System",
            "fontSize": 16,
          },
          Object {
            "alignSelf": "center",
            "fontFamily": "quicksand",
            "fontSize": 15,
            "margin": 5,
          },
        ]
      }
    >
      - or -
    </Text>
    <View
      accessible={true}
      focusable={true}
      forwardedRef={[Function]}
      onClick={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        Object {
          "alignItems": "center",
          "alignSelf": "center",
          "backgroundColor": "transparent",
          "borderBottomWidth": 1,
          "borderColor": "#5cb85c",
          "borderLeftWidth": 1,
          "borderRadius": 5,
          "borderRightWidth": 1,
          "borderTopWidth": 1,
          "borderWidth": 1,
          "color": "#2cb833",
          "elevation": null,
          "flexDirection": "row",
          "height": 45,
          "justifyContent": "space-between",
          "opacity": 1,
          "paddingBottom": 6,
          "paddingTop": 6,
          "shadowColor": null,
          "shadowOffset": null,
          "shadowOpacity": null,
          "shadowRadius": null,
        }
      }
    >
      <Text
        style={
          Array [
            Object {
              "backgroundColor": "transparent",
              "color": "#5cb85c",
              "fontFamily": "System",
              "fontSize": 16.5,
              "marginLeft": 0,
              "marginRight": 0,
              "paddingLeft": 16,
              "paddingRight": 16,
            },
            Object {
              "fontFamily": "quicksand",
              "fontSize": 15,
            },
          ]
        }
      >
        Browse Our Featured Charities
      </Text>
    </View>
  </View>
  <RCTScrollView
    collapsable={false}
    data={null}
    disableVirtualization={false}
    getItem={[Function]}
    getItemCount={[Function]}
    horizontal={false}
    initialNumToRender={10}
    keyExtractor={[Function]}
    maxToRenderPerBatch={10}
    onContentSizeChange={[Function]}
    onEndReachedThreshold={2}
    onGestureHandlerEvent={[Function]}
    onGestureHandlerStateChange={[Function]}
    onLayout={[Function]}
    onMomentumScrollEnd={[Function]}
    onScroll={[Function]}
    onScrollBeginDrag={[Function]}
    onScrollEndDrag={[Function]}
    removeClippedSubviews={false}
    renderItem={[Function]}
    renderScrollComponent={[Function]}
    scrollEventThrottle={50}
    stickyHeaderIndices={Array []}
    style={
      Object {
        "flex": 1,
        "width": "90%",
      }
    }
    updateCellsBatchingPeriod={50}
    viewabilityConfigCallbackPairs={Array []}
    windowSize={21}
  >
    <View />
  </RCTScrollView>
</RNCSafeAreaView>
`;
